python/lsst/daf/butler/core/timespan.py

```diff
@@ -78,6 +79,53 @@ class _SpecialTimespanBound(enum.Enum):
     Timespans that contain no points.
     """
 
+    @classmethod
+    def to_yaml(cls, dumper: yaml.Dumper, data: _SpecialTimespanBound) -> Any:
+        """Convert _SpecialTimespanBound into YAML format.
+
+        This produces a scalar node with a tag "!_SpecialTimespanBound" and
+        value being a name of _SpecialTimespanBound enum.
+
+        Parameters
+        ----------
+        dumper : `yaml.Dumper`
+            YAML dumper instance.
+        data : `_SpecialTimespanBound`
+            Data to be converted.
+        """
+        return dumper.represent_scalar(_SpecialTimespanBound_yaml_tag, data.name)
+
+    @classmethod
+    def from_yaml(cls, loader: yaml.SafeLoader, node: yaml.ScalarNode) -> Optional[_SpecialTimespanBound]:
+        """Convert YAML node into _SpecialTimespanBound.
+
+        Parameters
+        ----------
+        loader : `yaml.SafeLoader`
+            Instance of YAML loader class.
+        node : `yaml.ScalarNode`
+            YAML node.
+
+        Returns
+        -------
+        value : `_SpecialTimespanBound`
+            _SpecialTimespanBound instance, can be ``None``.
+        """
+        if node.value is not None:
+            return _SpecialTimespanBound[node.value]
+        return None
+
+
+# YAML tag name for _SpecialTimespanBound
+_SpecialTimespanBound_yaml_tag = "!lsst.daf.butler._SpecialTimespanBound"
```

Comment:
I do wonder if we want to have this tag burned into the YAML rather than the notion of EMPTY since the docstring for this class says we should always access it as `TimeSpan.EMPTY`.  Maybe @TallJimbo has thoughts.

Reply:
If we are allowed to make it whatever string we want, then yes, `Timespan.EMPTY` would be preferable.

If it's not difficult to override the YAML serialization to this extent, it'd arguably be even better to map the entire `Timespan` to a sentinal string value like "EMPTY", since it's really a property of the entire timespan rather than just one bound, but if it's a lot harder it's not worth the effort.

Do we know what the YAML looks like with this solution?

Reply:
The format before/after this change is on the JIRA here: https://jira.lsstcorp.org/browse/DM-32407?focusedCommentId=340215&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-340215

I was not sure that `_SpecialTimespanBound` is never going to have any other enum values for it, so I decided to keep it open for possible extensions. I was also thinking about replacing whole empty Timespan with a special YAML key, but this would mean adding more knowledge about Timespan details to YamlRepoExportBackend/YamlRepoImportBackend. Right now these two classes only know that Timespan has begin/end timestamps and it works OK for all cases so far. Making YAML representation more complicated adds more logic to those two classes. Alternatively we can define a special YAML representer for Timespan class itself to encapsulate that logic where it belongs, that adds YAML dependency to Timespan.

Reply:
Adding a YAML representer to `Timespan` does feel like the right solution.

Reply:
OK, I'll try something quick today on this ticket.